_posts/2024-08-14-running-a-hsm-realm.md

```diff
@@ -0,0 +1,221 @@
+particular specifications of the servers used and the number of HSMs in the
+realm.
+
+Local testing can be done with a single HSM, however test clusters should have a
+minimum of 3 HSMs in order to realistically exercise the consensus and failover
+operations. Production clusters should have at least 5 HSMs, as once the key
+ceremony has been performed additional HSMs can not be added to the realm. A 5
+HSM cluster gives some runway to handle failures.
+
+In our testing a 5 HSM cluster (with 3 replication groups) should be able to
+perform approximately 240 recover operations a second. This throughput drops off
+slightly as the size of the merkle tree increases. We observed a 5% drop in
+throughput going from a tree with 10M leaves to a tree with 500M leaves. Scaling
+to larger throughput will require more HSMs and replication groups.
+
+Improvements to overall availability can be achieved by running multiple
+independent realms with a N of M client configuration. e.g. 2 HSM realms and 1
+software realm with a 2 of 3 client configuration.
+
+## Google Cloud Platform
+
+The GCP services Bigtable, Secret Manager and Cloud Pub/Sub are used by the
+realm services. Credentials to access these will be needed for all hosts
+(regardless of role) in the cluster. Bigtable must be configured for a single
+zone as multi-zone configurations change the [semantics of write
+operations](https://cloud.google.com/bigtable/docs/replication-overview#consistency-model).
+
+## HSM Initialization
+
+For test clusters the `entrust_ops` tool that is part of the [HSM
+repo](https://github.com/juicebox-systems/juicebox-hsm-realm) can be used to
+initialize the NVRAM, create the Security World and the relevant realm keys. The
+[`entrust_hsm/README.md`](https://github.com/juicebox-systems/juicebox-hsm-realm/blob/main/entrust_hsm/README.md)
+has more information about this process. Unless you've purchased remote card
+readers physical access to the HSMs will be required during this process in
+order to connect the card reader and swap cards at the appropriate times.
+
+For production clusters a more rigorous key ceremony should be performed before
+the HSMs are installed in the servers. The key ceremony allows external parties
+to verify that the realm is running the expected HSM software and that only that
+version of the software has access to the keys. The
+[`ceremony`](https://github.com/juicebox-systems/ceremony) repo contains details
+of a possible key ceremony structure and process. The result of the ceremony
+includes a set of files called a Security World, a set of initialized HSMs and
+the signed software. The HSMs should then be installed in their production
+servers, and the security world files copied to the relevant location (typically
+/opt/nfast/kmdata/local) on each HSM host server.
+
+## Running the Agent
+
+On each HSM host (aka Agent) server the `entrust_agent` executable and the
```

Comment:
Can we explain how they get the `entrust_agent` executable, even if it's just some simple cargo commands? I think the more this can be hand holding, the more likely someone is to actually succeed at it

Reply:
I'll add a separate section on the build